File: java/exercises/practice/unit_test_c51974e3_1373_4477_9baa_4f8a92aef189/tx_coordinator/src/test/java/TxCoordinatorTest.java
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.RepeatedTest;
import org.junit.jupiter.api.Timeout;

import java.util.UUID;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.List;
import java.util.ArrayList;

import static org.junit.jupiter.api.Assertions.*;

// Dummy exceptions for simulation
class ServiceUnavailableException extends Exception {
    public ServiceUnavailableException(String message) {
        super(message);
    }
}

// The Service interface as defined in the problem statement
interface Service {
    boolean prepare(UUID transactionId) throws Exception;
    void commit(UUID transactionId) throws Exception;
    void rollback(UUID transactionId) throws Exception;
}

// Dummy implementation for success case
class AlwaysSuccessService implements Service {
    public boolean prepareCalled = false;
    public boolean commitCalled = false;
    public boolean rollbackCalled = false;

    @Override
    public boolean prepare(UUID transactionId) throws Exception {
        prepareCalled = true;
        return true;
    }

    @Override
    public void commit(UUID transactionId) throws Exception {
        commitCalled = true;
    }

    @Override
    public void rollback(UUID transactionId) throws Exception {
        rollbackCalled = true;
    }
}

// Dummy implementation for a service that fails during prepare by returning false
class AlwaysFailPrepareService implements Service {
    public boolean prepareCalled = false;
    public boolean rollbackCalled = false;

    @Override
    public boolean prepare(UUID transactionId) throws Exception {
        prepareCalled = true;
        return false;
    }

    @Override
    public void commit(UUID transactionId) throws Exception {
        // Should not be called if prepare fails
    }

    @Override
    public void rollback(UUID transactionId) throws Exception {
        rollbackCalled = true;
    }
}

// Dummy implementation for a service that throws exception during prepare
class ExceptionPrepareService implements Service {
    public boolean prepareCalled = false;
    public boolean rollbackCalled = false;

    @Override
    public boolean prepare(UUID transactionId) throws Exception {
        prepareCalled = true;
        throw new Exception("Prepare exception");
    }

    @Override
    public void commit(UUID transactionId) throws Exception {
        // Not expected
    }

    @Override
    public void rollback(UUID transactionId) throws Exception {
        rollbackCalled = true;
    }
}

// Dummy implementation for a service that simulates commit retries.
// It will throw exceptions a given number of times before succeeding.
class CommitRetryService implements Service {
    private final int failuresBeforeSuccess;
    private final AtomicInteger commitAttempts = new AtomicInteger(0);
    public boolean prepareCalled = false;
    public boolean commitCalled = false;
    public boolean rollbackCalled = false;

    public CommitRetryService(int failuresBeforeSuccess) {
        this.failuresBeforeSuccess = failuresBeforeSuccess;
    }

    @Override
    public boolean prepare(UUID transactionId) throws Exception {
        prepareCalled = true;
        return true;
    }

    @Override
    public void commit(UUID transactionId) throws Exception {
        int attempt = commitAttempts.incrementAndGet();
        if (attempt <= failuresBeforeSuccess) {
            throw new TimeoutException("Simulated timeout on commit attempt " + attempt);
        }
        commitCalled = true;
    }

    @Override
    public void rollback(UUID transactionId) throws Exception {
        rollbackCalled = true;
    }
}

// Dummy implementation for a service that always throws exception on rollback.
class RollbackExceptionService implements Service {
    public boolean prepareCalled = false;
    public boolean commitCalled = false;
    public boolean rollbackCalled = false;

    @Override
    public boolean prepare(UUID transactionId) throws Exception {
        prepareCalled = true;
        return false; // Force rollback scenario
    }

    @Override
    public void commit(UUID transactionId) throws Exception {
        commitCalled = true;
    }

    @Override
    public void rollback(UUID transactionId) throws Exception {
        rollbackCalled = true;
        throw new ServiceUnavailableException("Simulated failure in rollback");
    }
}

// Assume TxCoordinator implements the Distributed Transaction Coordinator functionality 
// according to the problem statement and is available in src/main/java.
// Its API includes: begin(), enlist(UUID transactionId, Service service),
// commit(UUID transactionId) which returns boolean, and rollback(UUID transactionId).
// For the purpose of testing, we assume its existence.

class TxCoordinator {
    // Internal storage for transactions: mapping transaction id to list of services
    private final ConcurrentMap<UUID, List<Service>> transactions = new ConcurrentHashMap<>();

    // Executor for retrying commit/rollback operations
    private final ScheduledExecutorService executor = Executors.newScheduledThreadPool(4);

    // Maximum number of retry attempts for commit/rollback operations
    private final int maxRetries = 3;
    // Delay between retries in milliseconds
    private final int retryDelayMs = 100;

    public UUID begin() {
        UUID txId = UUID.randomUUID();
        transactions.put(txId, new CopyOnWriteArrayList<>());
        return txId;
    }

    public void enlist(UUID transactionId, Service service) {
        List<Service> serviceList = transactions.get(transactionId);
        if (serviceList != null) {
            serviceList.add(service);
        } else {
            throw new IllegalArgumentException("Transaction does not exist");
        }
    }

    public boolean commit(UUID transactionId) {
        List<Service> serviceList = transactions.get(transactionId);
        if (serviceList == null) {
            throw new IllegalArgumentException("Transaction does not exist");
        }

        // Phase 1: Prepare phase
        for (Service service : serviceList) {
            try {
                boolean prepared = service.prepare(transactionId);
                if (!prepared) {
                    rollback(transactionId);
                    transactions.remove(transactionId);
                    return false;
                }
            } catch (Exception e) {
                rollback(transactionId);
                transactions.remove(transactionId);
                return false;
            }
        }

        // Phase 2: Commit phase with retries
        for (Service service : serviceList) {
            boolean committed = executeWithRetry(() -> {
                service.commit(transactionId);
                return true;
            });
            if (!committed) {
                rollback(transactionId);
                transactions.remove(transactionId);
                return false;
            }
        }
        transactions.remove(transactionId);
        return true;
    }

    public void rollback(UUID transactionId) {
        List<Service> serviceList = transactions.get(transactionId);
        if (serviceList == null) {
            throw new IllegalArgumentException("Transaction does not exist");
        }
        // Attempt to rollback all services with retries, logging error in case of failure.
        for (Service service : serviceList) {
            executeWithRetry(() -> {
                service.rollback(transactionId);
                return true;
            });
        }
        transactions.remove(transactionId);
    }

    // Helper method to execute an operation with retries on TimeoutException or ServiceUnavailableException
    private boolean executeWithRetry(Callable<Boolean> operation) {
        int attempts = 0;
        while (attempts < maxRetries) {
            try {
                return operation.call();
            } catch (TimeoutException | ServiceUnavailableException e) {
                attempts++;
                try {
                    Thread.sleep(retryDelayMs);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    return false;
                }
            } catch (Exception ex) {
                return false;
            }
        }
        return false;
    }
}

public class TxCoordinatorTest {

    private TxCoordinator coordinator;

    @BeforeEach
    public void setup() {
        coordinator = new TxCoordinator();
    }

    @Test
    public void testSuccessfulCommit() {
        UUID txId = coordinator.begin();
        AlwaysSuccessService service = new AlwaysSuccessService();
        coordinator.enlist(txId, service);

        boolean result = coordinator.commit(txId);
        assertTrue(result, "Transaction should commit successfully");
        assertTrue(service.prepareCalled, "Prepare should be called");
        assertTrue(service.commitCalled, "Commit should be called");
        assertFalse(service.rollbackCalled, "Rollback should not be called");
    }

    @Test
    public void testPrepareFailureTriggersRollback() {
        UUID txId = coordinator.begin();
        AlwaysFailPrepareService service = new AlwaysFailPrepareService();
        coordinator.enlist(txId, service);

        boolean result = coordinator.commit(txId);
        assertFalse(result, "Transaction should fail to commit due to prepare failure");
        assertTrue(service.prepareCalled, "Prepare should be called");
        assertTrue(service.rollbackCalled, "Rollback should be called after prepare failure");
    }

    @Test
    public void testExceptionDuringPrepareTriggersRollback() {
        UUID txId = coordinator.begin();
        ExceptionPrepareService service = new ExceptionPrepareService();
        coordinator.enlist(txId, service);

        boolean result = coordinator.commit(txId);
        assertFalse(result, "Transaction should fail to commit due to exception in prepare");
        assertTrue(service.prepareCalled, "Prepare should be called");
        assertTrue(service.rollbackCalled, "Rollback should be called after exception");
    }

    @Test
    public void testCommitWithRetrySuccess() {
        UUID txId = coordinator.begin();
        // Service will fail commit twice before succeeding on the third attempt.
        CommitRetryService service = new CommitRetryService(2);
        coordinator.enlist(txId, service);

        boolean result = coordinator.commit(txId);
        assertTrue(result, "Transaction should eventually commit successfully after retries");
        assertTrue(service.prepareCalled, "Prepare should be called");
        assertTrue(service.commitCalled, "Commit should ultimately succeed");
    }

    @Test
    public void testRollbackWithException() {
        UUID txId = coordinator.begin();
        RollbackExceptionService service = new RollbackExceptionService();
        coordinator.enlist(txId, service);

        // Commit should fail because prepare returns false, triggering rollback.
        boolean result = coordinator.commit(txId);
        assertFalse(result, "Transaction should fail commit due to prepare failure");
        assertTrue(service.prepareCalled, "Prepare should be called");
        // Even though rollback throws exception, our coordinator retries and eventually gives up.
        assertTrue(service.rollbackCalled, "Rollback should have been attempted");
    }

    @Test
    @Timeout(10)
    public void testConcurrentTransactions() throws InterruptedException, ExecutionException {
        int numThreads = 10;
        int transactionsPerThread = 20;
        ExecutorService exec = Executors.newFixedThreadPool(numThreads);
        List<Callable<Boolean>> tasks = new ArrayList<>();

        for (int i = 0; i < numThreads; i++) {
            tasks.add(() -> {
                boolean allSuccess = true;
                for (int j = 0; j < transactionsPerThread; j++) {
                    UUID txId = coordinator.begin();
                    AlwaysSuccessService service1 = new AlwaysSuccessService();
                    CommitRetryService service2 = new CommitRetryService(1);
                    coordinator.enlist(txId, service1);
                    coordinator.enlist(txId, service2);
                    boolean result = coordinator.commit(txId);
                    if (!result) {
                        allSuccess = false;
                    }
                }
                return allSuccess;
            });
        }

        List<Future<Boolean>> futures = exec.invokeAll(tasks);
        exec.shutdown();
        for (Future<Boolean> future : futures) {
            assertTrue(future.get(), "All concurrent transactions should commit successfully");
        }
    }
}

File: cpp/exercises/practice/unit_test_45cb4879_ea27_46e4_a6f8_68ff96c10928/tx_coordinator/tx_coordinator_test.cpp
#include "tx_coordinator.h"
#include <vector>
#include <thread>
#include <atomic>
#include <chrono>
#include "catch.hpp"

using namespace std;

class MockService : public Service {
public:
    MockService(int id, bool prepareSuccess, bool commitSuccess, bool rollbackSuccess)
        : id_(id), prepareSuccess_(prepareSuccess), 
          commitSuccess_(commitSuccess), rollbackSuccess_(rollbackSuccess) {}

    bool Prepare() override {
        return prepareSuccess_;
    }

    bool Commit() override {
        return commitSuccess_;
    }

    bool Rollback() override {
        return rollbackSuccess_;
    }

    int GetId() override {
        return id_;
    }

private:
    int id_;
    bool prepareSuccess_;
    bool commitSuccess_;
    bool rollbackSuccess_;
};

TEST_CASE("Single service successful transaction") {
    TransactionCoordinator coordinator;
    auto service = make_shared<MockService>(1, true, true, true);
    coordinator.RegisterService(service.get());
    
    vector<Service*> services = {service.get()};
    REQUIRE(coordinator.BeginTransaction(services) == true);
}

TEST_CASE("Single service failed prepare") {
    TransactionCoordinator coordinator;
    auto service = make_shared<MockService>(1, false, true, true);
    coordinator.RegisterService(service.get());
    
    vector<Service*> services = {service.get()};
    REQUIRE(coordinator.BeginTransaction(services) == false);
}

TEST_CASE("Multiple services all successful") {
    TransactionCoordinator coordinator;
    auto service1 = make_shared<MockService>(1, true, true, true);
    auto service2 = make_shared<MockService>(2, true, true, true);
    coordinator.RegisterService(service1.get());
    coordinator.RegisterService(service2.get());
    
    vector<Service*> services = {service1.get(), service2.get()};
    REQUIRE(coordinator.BeginTransaction(services) == true);
}

TEST_CASE("Multiple services one fails prepare") {
    TransactionCoordinator coordinator;
    auto service1 = make_shared<MockService>(1, true, true, true);
    auto service2 = make_shared<MockService>(2, false, true, true);
    coordinator.RegisterService(service1.get());
    coordinator.RegisterService(service2.get());
    
    vector<Service*> services = {service1.get(), service2.get()};
    REQUIRE(coordinator.BeginTransaction(services) == false);
}

TEST_CASE("Concurrent transactions") {
    TransactionCoordinator coordinator;
    auto service1 = make_shared<MockService>(1, true, true, true);
    auto service2 = make_shared<MockService>(2, true, true, true);
    coordinator.RegisterService(service1.get());
    coordinator.RegisterService(service2.get());
    
    atomic<int> successCount(0);
    vector<thread> threads;
    
    for (int i = 0; i < 10; ++i) {
        threads.emplace_back([&]() {
            vector<Service*> services = {service1.get(), service2.get()};
            if (coordinator.BeginTransaction(services)) {
                successCount++;
            }
        });
    }
    
    for (auto& t : threads) {
        t.join();
    }
    
    REQUIRE(successCount == 10);
}

TEST_CASE("Service timeout during prepare") {
    TransactionCoordinator coordinator;
    auto service = make_shared<MockService>(1, false, false, false);
    coordinator.RegisterService(service.get());
    
    vector<Service*> services = {service.get()};
    REQUIRE(coordinator.BeginTransaction(services) == false);
}

TEST_CASE("Unregistered service") {
    TransactionCoordinator coordinator;
    auto service = make_shared<MockService>(1, true, true, true);
    
    vector<Service*> services = {service.get()};
    REQUIRE_THROWS_AS(coordinator.BeginTransaction(services), std::runtime_error);
}

TEST_CASE("Service failure during commit") {
    TransactionCoordinator coordinator;
    auto service1 = make_shared<MockService>(1, true, false, true);
    auto service2 = make_shared<MockService>(2, true, true, true);
    coordinator.RegisterService(service1.get());
    coordinator.RegisterService(service2.get());
    
    vector<Service*> services = {service1.get(), service2.get()};
    REQUIRE(coordinator.BeginTransaction(services) == false);
}

TEST_CASE("Service failure during rollback") {
    TransactionCoordinator coordinator;
    auto service1 = make_shared<MockService>(1, false, true, false);
    auto service2 = make_shared<MockService>(2, true, true, true);
    coordinator.RegisterService(service1.get());
    coordinator.RegisterService(service2.get());
    
    vector<Service*> services = {service1.get(), service2.get()};
    REQUIRE(coordinator.BeginTransaction(services) == false);
}

TEST_CASE("Mixed success/failure scenarios") {
    TransactionCoordinator coordinator;
    auto service1 = make_shared<MockService>(1, true, true, true);
    auto service2 = make_shared<MockService>(2, false, true, false);
    auto service3 = make_shared<MockService>(3, true, false, true);
    coordinator.RegisterService(service1.get());
    coordinator.RegisterService(service2.get());
    coordinator.RegisterService(service3.get());
    
    vector<Service*> services = {service1.get(), service2.get(), service3.get()};
    REQUIRE(coordinator.BeginTransaction(services) == false);
}